Replace e2essh on e2e service tests

JornShen opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind bug
/kind cleanup
 
**What this PR does / why we need it**:

- Backgroud

There are a few e2e sig-network tests that are using e2essh, limiting the number of environments where they can run.
some of them can be easily replaced executing the commands on a pod. This PR can reach that e2e network service test can execute whever hosts have ssh or not.

ref: https://github.com/kubernetes/kubernetes/issues/94703#issue-698508813

- What's the changes 

There are three tests  in network service suits of e2e relevant to it.

> ginkgo.It("should be able to up and down services", ...)
> ginkgo.It("should implement service.kubernetes.io/service-proxy-name",  ...)
> ginkgo.It("should implement service.kubernetes.io/headless", ...)

change:  replace ssh host to exec host network pod to verify the service up

- Running Result
```
{"msg":"PASSED [sig-network] Services should implement service.kubernetes.io/service-proxy-name","total":1,"completed":1,"skipped":3969,"failed":0}

{"msg":"PASSED [sig-network] Services should be able to up and down services","total":1,"completed":1,"skipped":818,"failed":0}

{"msg":"PASSED [sig-network] Services should implement service.kubernetes.io/headless","total":1,"completed":1,"skipped":3288,"failed":0}

```


**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/94703

**release-note**
```release-note
NONE
```

JornShen commented:
/cc @aojea @Lion-Wei

Lion-Wei commented:
/ok-to-test
/assign

aojea commented:
let's have buy in from sig-testing too
/assign @BenTheElder @spiffxp 
the improvement is those test run in KIND too now  ;)

aojea commented on test/e2e/network/service.go:
```diff
@@ -1166,10 +1205,9 @@ var _ = SIGDescribe("Services", func() {
 	ginkgo.It("should work after restarting apiserver [Disruptive]", func() {
 		// TODO: use the ServiceTestJig here
 		e2eskipper.SkipUnlessProviderIs("gce", "gke")
```
is this a gce test only?

JornShen replied:
yes, same as test case ("should work after restarting kube-proxy [Disruptive]").  I try to remove this skip, but no success. Beacause I find that  next the test restart api-server or kube-proxy  will still judge  the env is gce or gke. 

https://github.com/kubernetes/kubernetes/blob/fcba5fe4e4eb1b1a2e2b14d2a68037fecd75e3ab/test/e2e/network/service.go#L712-L716

and

https://github.com/kubernetes/kubernetes/blob/fcba5fe4e4eb1b1a2e2b14d2a68037fecd75e3ab/test/e2e/network/service.go#L3802-L3807

aojea replied:
and what happens if you remove the check for `ProviderIs` in those functions? does it work? can you add a temp commit to this PR removing it to see the result?
I mean, the test is tagged as disruptive, people can filter based on the tags, and the idea is to remove all the provider specific code ...

JornShen replied:
All right. I think these two test case can be discussed about. Before I say that that no success, it was running in my local robot using local-cluster-up.sh to simulate a k8s cluster. But when I see the code,  I find that the code treat kube-apiserver and kube-proxy as container pod which is different from the process in my local env. So the test will failed in my env definitely.

Lion-Wei commented on test/e2e/network/service.go:
```diff
@@ -1080,12 +1130,7 @@ var _ = SIGDescribe("Services", func() {
-			framework.Failf("No ssh-able nodes")
-		}
-		host := hosts[0]
+		host := getSSHHost(cs)
```
Move `checkHostSupportSSH` here would change the current situation that this case can only run in providers gce/gke/aws/local.

JornShen replied:
I will remove it later. This case will not exist any host var.

Lion-Wei commented on test/e2e/network/service.go:
```diff
@@ -303,11 +303,11 @@ func StopServeHostnameService(clientset clientset.Interface, ns, name string) er
-// names are compared with the given expectedPods list after a sort | uniq.
+// given host or from within a pod(if there are no host satisfied, we will only verify from pod).
+// The host is expected to be an SSH-able node in the cluster. Each pod in the service is expected to echo its name.
+// These names are compared with the given expectedPods list after a sort | uniq.
```
IMO, we still need check hoth host network and contianer network, our target is to use `framework.RunHostCmd` replace `e2essh.SSH`.

JornShen replied:
You are right. I should think it more carefully. 👍

Lion-Wei commented on test/e2e/network/service.go:
```diff
@@ -319,9 +319,11 @@ func verifyServeHostnameServiceUp(c clientset.Interface, ns, host string, expect
-	commands := []func() string{
+
+	var commands []func() string
+	if host != "" {
```
We still need verify host network connectivity, but this parameter is not necessary anymore, you can launch a pod with hostNetwork, and use `framework.RunHostCmd` to run command from this pod, reference:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/network/service.go#L2007

JornShen replied:
done.

Lion-Wei commented on test/e2e/network/service.go:
```diff
@@ -330,20 +332,21 @@ func verifyServeHostnameServiceUp(c clientset.Interface, ns, host string, expect
 	}
 
+	// verify service from pod
+	commands = append(commands, func() string {
```
IMO, the original build commands code would be better. 🤔

Lion-Wei commented on test/e2e/network/service.go:
```diff
@@ -384,25 +387,49 @@ func verifyServeHostnameServiceUp(c clientset.Interface, ns, host string, expect
-func verifyServeHostnameServiceDown(c clientset.Interface, host string, serviceIP string, servicePort int) error {
+// Firstly it will verify from the host. If there are no host satisfied or verify from host no success,
+//it will verify from the exec pod then.
+func verifyServeHostnameServiceDown(c clientset.Interface, ns, host string, serviceIP string, servicePort int) error {
```
Same with `verifyServeHostnameServiceUp`

JornShen replied:
done.

JornShen commented:
/cc @Lion-Wei @aojea

aojea commented on test/e2e/network/service.go:
```diff
@@ -3798,84 +3703,64 @@ func validateEndpointsPorts(c clientset.Interface, namespace, serviceName string
-	v, err := cs.Discovery().ServerVersion()
+// restartComponent restarts component static pod
+func restartComponent(cs clientset.Interface, cName, ns string, labelSet labels.Set) error {
+	pods, err := getComponentPod(cs, ns, labelSet)
```
this is assuming that the apiserver or kubeproxy are pods, and that is not always true

aojea replied:
however, seems the test was already considering that, right?

aojea replied:
the `sshRestartMaster()` is not clear about that :/

aojea replied:
also the `restartKubeProxy` kills a process

JornShen replied:
It seems that the case before thinking the api-server or kubelet as static pod. So they kill the pods to let kubelet restart the component (kube-proxy or apiserver). As you say, there is no doubt that apiserver or kube-proxy can be running as process. But if we want to restart them, who will restart them when we kill them or other way to stop them? I used to running the local-up-cluster.sh as they are process, no will start them when i kill them.  :）

aojea replied:
yeah, I'm sorry, seems this is not a big win and overcomplicates everything too much

BenTheElder replied:
An init system will restart them?

aojea replied:
let's tackle the Disruptive tests in another PR, and have tese tests now working on KIND jobs
```
ginkgo.It("should be able to up and down services", ...)
ginkgo.It("should implement service.kubernetes.io/service-proxy-name", ...)
ginkgo.It("should implement service.kubernetes.io/headless", ...)
```

JornShen replied:
All right. I will take annother PR to see more details about the change:
```
{"msg":"PASSED [sig-network] Services should work after restarting apiserver [Disruptive]","total":1,"completed":1,"skipped":2981,"failed":0}

{"msg":"PASSED [sig-network] Services should work after restarting kube-proxy [Disruptive]","total":1,"completed":1,"skipped":2061,"failed":0}
```

aojea commented on test/e2e/network/service.go:
```diff
@@ -83,6 +82,11 @@ const (
 	// affinity is enabled.
 	AffinityConfirmCount = 15
+
+	kubeProxyLabelName  = "kube-proxy"
+	kubeAPIServerLabelName = "kube-apiserver"
+	clusterComponentKey = "component"
+	clusterAddonLabelKey  = "k8s-app"
```
are these labels standard?
can you add a comment explaining the origin?

by the way, the go-fmt on this file, it seems need to be indented

JornShen replied:
the const str "kube-apiserver"  and "component" is appear before. And I just set them as const var in this PR. For go-fmt I will fix it right now. And also the comment explaining.

aojea commented:
@JornShen I have to apologize for this comment https://github.com/kubernetes/kubernetes/pull/94822#discussion_r489408728, I only wanted to check if it was feasible and simple to change those tests, but it seems that it made a lot of assumptions on how the kube-porxy and apiserver and deployed. Based on that I think we should forget about those `[Disruptive]` tests and the new logic about restarting service, the rest LGTM overall

JornShen committed:
Replace e2essh on e2e service tests

test can execute whever hosts have ssh or not

relevant case:

"should be able to up and down services"
"should implement service.kubernetes.io/service-proxy-name"
"should implement service.kubernetes.io/headless"

aojea commented:
/lgtm
thanks

Lion-Wei commented:
/lgtm
thanks

dcbw commented:
/approve

spiffxp commented:
/retest

spiffxp reviewed (commented):
/hold
I just want to confirm this isn't adversely impacting conformance tests, need to read through

spiffxp commented:
/hold cancel
/lgtm
thanks!
